### PR TITLE
Use fmt::format in NCCLUtils and ProcessGroupNCCL instead of c10::str

### LINKED ISSUE #64604: Replace usages of c10::str with fmt::format in distributed
## 🐛 Bug

As per the discussion in https://github.com/pytorch/pytorch/pull/64241, we would prefer to use `fmt` library (specifically `fmt::format`) to format strings/error messages instead of c10::str. However the `fmt` library is currently only linked to libtorch_python sources, see the PR that added it: https://github.com/pytorch/pytorch/pull/64241. 

To use `fmt::format` we would probably need to link fmt against libtorch, and then we can update the callsites introduced in `ProcessGroupNCCL` and `NCCLUtils`  in the above PR.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

### PR DESCRIPTION
**Pull request by andreigh**
Fixes #64604

### CODE REVIEW

File: cmake/Dependencies.cmake [hunk 1]
```diff
@@ -1346,7 +1346,8 @@ if(USE_NCCL)
     caffe2_update_option(USE_NCCL OFF)
   elseif(USE_CUDA)
     include(${CMAKE_CURRENT_LIST_DIR}/External/nccl.cmake)
-    list(APPEND Caffe2_CUDA_DEPENDENCY_LIBS __caffe2_nccl)
+    # fmt::fmt-header-only is needed by the NCCL code in torch/csrc/distributed/c10d 
+    list(APPEND Caffe2_CUDA_DEPENDENCY_LIBS __caffe2_nccl fmt::fmt-header-only)
```
Comment (andreigh) [inline cmake/Dependencies.cmake:1350]: Not very sure this is the best way to link fmt::fmt-header-only)

**Comment by andreigh** [comment]
@rohan-varma any chance you can take a quick look ? Thank you !

**Comment by andreigh** [comment]
@rohan-varma gentle ping, thank you !

**Review by fduwjj — APPROVED** [review]
LGTM. Can we rebase to handle the conflict?

**Comment by jansel** [comment]
@andreigh this PR causes the PyTorch build to fail on my Ubuntu 23.04 machine with:
```
FAILED: caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp.o 
/usr/bin/ccache /usr/bin/g++-10 -DAT_PER_OPERATOR_HEADERS -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTORCH_CUDA_BUILD_MAIN_LIB -DUSE_C10D_GLOO -DUSE_CUDA -DUSE_DISTRIBUTED -DUSE_EXPERIMENTAL_CUDNN_V8_API -DUSE_EXTERNAL_MZCRC -DUSE_FLASH_ATTENTION -DUSE_MEM_EFF_ATTENTION -DUSE_RPC -DUSE_TENSORPIPE -D_FILE_OFFSET_BITS=64 -Dtorch_cuda_EXPORTS -I/home/jansel/pytorch/build/aten/src -I/home/jansel/pytorch/aten/src -I/home/jansel/pytorch/build -I/home/jansel/pytorch -I/home/jansel/pytorch/cmake/../third_party/benchmark/include -I/home/jansel/pytorch/third_party/onnx -I/home/jansel/pytorch/build/third_party/onnx -I/home/jansel/pytorch/third_party/foxi -I/home/jansel/pytorch/build/third_party/foxi -I/home/jansel/pytorch/aten/src/THC -I/home/jansel/pytorch/aten/src/ATen/cuda -I/home/jansel/pytorch/aten/src/ATen/../../../third_party/cutlass/include -I/home/jansel/pytorch/build/caffe2/aten/src -I/home/jansel/pytorch/aten/src/ATen/.. -I/home/jansel/pytorch/c10/cuda/../.. -I/home/jansel/pytorch/c10/.. -I/home/jansel/pytorch/third_party/tensorpipe -I/home/jansel/pytorch/build/third_party/tensorpipe -I/home/jansel/pytorch/third_party/tensorpipe/third_party/libnop/include -I/home/jansel/pytorch/torch/csrc/api -I/home/jansel/pytorch/torch/csrc/api/include -isystem /home/jansel/pytorch/build/third_party/gloo -isystem /home/jansel/pytorch/cmake/../third_party/gloo -isystem /home/jansel/pytorch/cmake/../third_party/tensorpipe/third_party/libuv/include -isystem /home/jansel/pytorch/cmake/../third_party/googletest/googlemock/include -isystem /home/jansel/pytorch/cmake/../third_party/googletest/googletest/include -isystem /home/jansel/pytorch/third_party/protobuf/src -isystem /home/jansel/conda/envs/pytorch/include -isystem /home/jansel/pytorch/third_party/gemmlowp -isystem /home/jansel/pytorch/third_party/neon2sse -isystem /home/jansel/pytorch/third_party/XNNPACK/include -isystem /home/jansel/pytorch/third_party/ittapi/include -isystem /home/jansel/pytorch/cmake/../third_party/eigen -isystem /usr/local/cuda/include -isystem /home/jansel/pytorch/third_party/ideep/mkl-dnn/include/oneapi/dnnl -isystem /home/jansel/pytorch/third_party/ideep/include -isystem /home/jansel/pytorch/cmake/../third_party/cudnn_frontend/include -D_GLIBCXX_USE_CXX11_ABI=1 -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOROCTRACER -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=old-style-cast -Wno-missing-braces -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow -DHAVE_AVX512_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG -std=gnu++17 -fPIC -DMKL_HAS_SBGEMM -DTORCH_USE_LIBUV -DCAFFE2_USE_GLOO -Wall -Wextra -Wdeprecated -Wno-unused-parameter -Wno-unused-function -Wno-missing-field-initializers -Wno-unknown-pragmas -Wno-type-limits -Wno-array-bounds -Wno-strict-overflow -Wno-strict-aliasing -Wno-maybe-uninitialized -fvisibility=hidden -O2 -MD -MT caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp.o -MF caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp.o.d -o caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp.o -c /home/jansel/pytorch/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
/home/jansel/pytorch/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1:10: fatal error: fmt/format.h: No such file or directory
    1 | #include <fmt/format.h>
      |          ^~~~~~~~~~~~~~
compilation terminated.
```

**Comment by cyyever** [comment]
The header inclusions above  '#ifdef USE_C10D_NCCL' should be put inside '#ifdef USE_C10D_NCCL' in ProcessGroupNCCL.cpp

**Comment by andreigh** [comment]
Prepared PR #112112, however I can't find any nccl targets in any .bzl file nor buck ?

**Comment by fduwjj** [comment]
@andreigh The BUCK file is meta internal.. So if there are some internal changes (if needed), shall I or any Meta employee land the diff instead of merging PR here. @malfet

**Pull request closed without merging**